src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,209 @@
+            throws StorageQueryException, TotpNotEnabledException, UnknownDeviceException, InvalidTotpException {
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // Find the device:
+        TOTPDevice matchingDevice = null;
+        for (TOTPDevice device : devices) {
+            if (device.deviceName.equals(deviceName)) {
+                if (device.verified) {
+                    // TODO: Should we just return a boolean here?
+                    return new VerifyDeviceResponse(true);
+                } else {
+                    matchingDevice = device;
+                    break;
+                }
+            }
+        }
+        if (matchingDevice == null) {
+            throw new UnknownDeviceException();
+        }
+
+        // // Insert the code into the list of used codes:
+        TOTPUsedCode newCode = new TOTPUsedCode(userId, code, true, System.currentTimeMillis() + 1000 * 60 * 5);
+        totpStorage.insertUsedCode(newCode);
+
+        // Check if the code is valid:
+        if (!checkCode(matchingDevice, code)) {
+            throw new InvalidTotpException();
+        }
+
+        // Check if the code is unused:
+        TOTPUsedCode[] usedCodes = totpStorage.getUsedCodes(userId);
+        for (TOTPUsedCode usedCode : usedCodes) {
+            if (usedCode.code.equals(code) && usedCode.isValidCode) {
+                throw new InvalidTotpException();
+            }
+        }
+
+        totpStorage.markDeviceAsVerified(userId, deviceName);
+        return new VerifyDeviceResponse(false);
+    }
+
+    public static void verifyCode(Main main, String userId, String code, boolean allowUnverifiedDevices)
+            throws StorageQueryException, TotpNotEnabledException, InvalidTotpException,
+            LimitReachedException {
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        // Check if the user has any devices:
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // FIXME: Every unused_code should be linked to the device it was used for.
+        // Otherwise, if a user has multiple devices, and they use the same code for
+        // both,
+        // then the code will be considered as used for both devices. This could cause
+        // UX
+        // issues.
+        // If we do this, then it also means that we need to assign a device ID to each
+        // device (OR use
+        // (userId, deviceName) as the ID)
+
+        // Check if the code has been successfully used by the user (for any device):
+        TOTPUsedCode[] usedCodes = totpStorage.getUsedCodes(userId);
```

Comment:
```suggestion
        TOTPUsedCode[] usedCodes = totpStorage.getNonExpiredUsedCodesInAscendingOrderByTime(userId);
```

Reply:
That's too lengthly. Named it as `getNonExpiredUsedCodes()` and mentioned `DESC` order in Javadocs for this func
Also modified the TOTPQueries method, it is now called `getNonExpiredUsedCodesDescOrder()`. 

Also, using descending order for simpler code for checking latest 3 attempts..